### PR TITLE
🦄  Utilize glob ignore option for massive perf win 🦄

### PR DESCRIPTION
**Pull request by samccone**
#### TL;DR;

By passing the ignore options directly to glob instead of filtering the files inside of eslint we see a *significant* perf boost.

-----

#### eslint perf walk through

[image](https://youtu.be/iMqi55rcR00?t=1770)

This perf analysis is using a version of node built off of this patch
https://github.com/nodejs/node/pull/6792

-------

First off, **thanks SO much eslint**, it is great and I ❤️  it. So onto the rational and story behind this PR. I was hacking away on a project https://github.com/GoogleChrome/lighthouse and I noticed that running lint was taking... multiple seconds (like 7). This seemed kinda long given the relatively small size of our project .. and given the fact that the project has an _eslintignore_ file I was really confused as to what was taking so long.

Before this change, running eslint over a normal (non-large) node.js project yielded a very interesting perf profile...

image

Now when I switched to using the ignore option in glob the results then looked like this:

image

Understanding what the code is doing can be kindy tricky here so, I have created two images to illustrate the impact of this change to eslint.

### Before:

image

### After: 
image

With this change applied, it roughly dropped the total run time on our project from 8 seconds to 4 seconds... or a `~50%` improvement in our lint. awesome! ✨  ✨  ✨  

--

While this might not be the perfect final solution I think that it is a good place to start :) Let me know if there is any more info that I can provide. Thanks! 🎅

### CODE REVIEW

**Comment by platinumazure** [comment]
On top of the concerns raised by @eslintbot and @jquerybot, I can say that we'll definitely want to see some unit tests and possibly some functional tests, so we don't regress here. :smile: 

Oh, and here's an issue you might want to track: #5679

Thanks for contributing!

**Comment by nzakas** [comment]
@samccone thanks for this. Some questions for you:

1. Can you explain what the change is doing? It looks like maybe you're passing regexs to ignore?
2. Do you know how this would affect re-including previously ignored files using `!` at the start of the pattern?
3. Our ignore files follow gitignore syntax, but I think glob follows just minimatch. I'd that a concern here?
4. Do you intend to work with us to land this? (If not, no worries, we still appreciate the effort and will apply your learnings on #5679.)

**Comment by samccone** [comment]
4) Yep I will gladly work with y'all to land this ... just give me a few days :)

**Comment by hzoo** [comment]
Oh ok, this is primarily for the case when you use `"lint": "eslint ."` as your command as opposed to specifying your folders explicitly (as mentioned in https://github.com/eslint/eslint/issues/5679)?

**Comment by gronke** [comment]
I like that approach, but share the concerns that it could interfere with the gitignore syntax. 

Looking into the [ignore section of the glob package](https://github.com/isaacs/node-glob/blob/68f2509d237babec64cf03adf72531127fd80339/common.js#L219-L235) shows how easy it would be to also use [node-ignore](https://github.com/kaelzhang/node-ignore) instead of minmatch there.

**Comment by samccone** [comment]
I think that 1 possible solution @nzakas is to write a transformation that turns the gitignore style ignores into an ignore format that glob understands.. this transform could happen in the dumpJoinedResults fn.

The advantage here is that it would not be breaking, and would require no upstream changes to glob. 

Thoughts? ✨

**Comment by samccone** [comment]
hi @eslintbot I do have (fixes #....) 🐐

**Comment by gyandeeps** [comment]
@samccone Since you dont have fixes on the first line that why bot is not happy. Hope this helps. :)

**Comment by ilyavolodin** [comment]
@samccone transforming gitignore to minmatch patterns would work, I'm just not sure how easy it would be, due to the fact that gitignore allows un-ignoring files/directories, and I don't think minmatch supports that.
Patch that @gronke linked looks interesting. If it would get merged into node-glob, then no more work needs to be done on this PR, and everything should just work.

**Comment by gronke** [comment]
The mapper between gitignore and minmatch syntax sounds like a project on it's own. Preferably we can land the pull-request in the glob package and use it accordingly. We could remove a bit more code with this PR that becomes obsolete with filtering files upfront.

**Comment by samccone** [comment]
@ilyavolodin @gronke ... I agree the upstream patch would be awesome :) I am just thinking out loud of possible ways we can handle this in case the patch does not land.

As far as the mapping lib, I also agree it does seem like a project on its own :) 🐔 
As far as the unsupported gitignore style `=>` glob conversion for unsupported formats, I figure that we would just drop those from the ignore glob and fall back to the manual filter logic that we already have.

**Comment by nzakas** [comment]
Back on #5679, what we were talking about is just passing `node_modules/` and `bower_components/` as `ignore` to `glob.sync()` whenever either a) there was no `.eslintignore` file or b) there is an `.eslintignore` file but there is no pattern that uses `!` (which is the edge case).

Doing that means we don't have to worry about `.gitignore` patterns vs. minimatch patterns. And since the major cause of concern is `node_modules` and `bower_components` directory traversal, we get a lot of the benefit without much downside.

**Comment by gronke** [comment]
@nzakas what you suggest is probably the quickest fix for #5679.

Some thoughts why filtering within glob would still be beneficial:
- For the ignore feature the location of `node_modules/` and `bower_components/` depends on the location of the ignore file. (The delta between the ignore file location and the current work directory should be added as prefix to each relative line of the ignore file when passed to glob)
- Differing directory names for npm or bower packages do not have an impact on performance
- We will ignore the filtered files anyway for further processing. This is the earliest possible step to get rid of them.

**Comment by nzakas** [comment]
> For the ignore feature the location of node_modules/ and bower_components/ depends on the location of the ignore file. (The delta between the ignore file location and the current work directory should be added as prefix to each relative line of the ignore file when passed to glob)

Yes, of course. That's an implementation detail - I was speaking generally. 

> Differing directory names for npm or bower packages do not have an impact on performance

To do this, we would have to go through the trouble of converting `.gitignore` patterns into minimatch patterns. As I said above, I don't think it's worth that headache right now (it took us forever to get ignores working correctly). We can always come back to that in the future if necessary, but I'm most interested in getting this perf boost for the majority of users rather than waiting to get something a lot more complicated that works for every edge case. 

> We will ignore the filtered files anyway for further processing. This is the earliest possible step to get rid of them.

We can only do that if no patterns contain `!`, otherwise, we can't re-include.

**Comment by alberto** [comment]
Besides not being the proper fix for the reasons mentioned above, I think is isn't even working.

I'm working in a proper fix for this.

**Comment by gronke** [comment]
> To do this, we would have to go through the trouble of converting .gitignore patterns into minimatch patterns.

https://github.com/isaacs/node-glob/pull/260 makes glob use node-ignore. With that change, we're able to pass in the same ignore patterns built by the `ignored-paths` library directly to glob.

> We can only do that if no patterns contain !, otherwise, we can't re-include.

.. unless glob uses the same filter framework as we do in a later stage (node-ignore or another .gitignore compatible one)

**Comment by nzakas** [comment]
@gronke thanks, we'll keep an eye on that PR. I think in the meantime, we should just see if we can get the common case working to get most of the benefit and we can always followup to make things incrementally better later on.

**Comment by samccone** [comment]
@nzakas I am totally cool with just updating this PR to pass node_modules and bower_components and call this "done" until we have a better path forward. Just let me know. :)

**Comment by nzakas** [comment]
@samccone it looks like @alberto did just that in #6276

**Comment by samccone** [comment]
#6276 👍 sounds great -- I think there is still value in figuring out how we can utilize glob's ignore option for real though since while node_modules and bower_components are big offenders there are many other permutations of this issue.

I see several options.

1. Convert gitignore pattern to glob matches (figuring out how we will handle the negation case) on the fly
2. Land the patch in glob (unlikely) 
3. Migrate eslint to use glob style ignores (imho this is actually a very reasonable solution since glob ignores are much more common in the node ecosystem as compared to gitignore patterns)
4... maybe other options that I can not think of.

Thoughts?

**Comment by ilyavolodin** [comment]
@samccone re 3. We just switched eslintignore to use gitignore style from minimatch, since we were getting a lot of issues with people using gitignore patterns and having unexpected results. I don't think we can switch back to minimatch at this point. But options 1 or 2 would work, I think. 1 might be complicated, because I don't think minimatch covers everything that node-ignore covers. Number 2 would be ideal, but since there hasn't been any movement at all on the PR that @gronke posted, it does seem unlikely.

**Comment by samccone** [comment]
For option 1... even if it was not 100% and was only partial we still have the second pass ignore work in eslint.. the goal for 1 would be to cover enough to get massive wins for the majority of people :)

If we can get consensus on even a simply naive but correct implementation of **1** I can take a pass at it.

**Comment by alberto** [comment]
I made #6276 with the intention that it could be extended to include more ignored dirs if we can ensure they are ignored. I think it's doable in some cases, if we want to take that route.

**Comment by samccone** [comment]
>  it could be extended to include more ignored dirs if we can ensure they are ignored.

I want to ensure that we can give users the eslintignore => glob ignore "native" option. that will work in "most" cases given :)

**Comment by samccone** [comment]
As #6276 has now landed I will close this off for now. 

I still think option 1 listed about would be useful :)
